### PR TITLE
Add Power distribution to Generator

### PR DESCRIPTION
**Pull request by povinsahu1909**
This PR adds power distribution to Generator.

Following is the performance comparison on GTX1050.

size(nrows * ncols) | numpy (sec) | cupyx (sec)
--- | --- | ---
10000 * 1 | 0.000 | 0.000
12000 * 1 | 0.001 | 0.000
15000 * 1 | 0.001 | 0.000
20000 * 1 | 0.001 | 0.000
10000 * 1000 | 0.409 | 0.050
12000 * 1000 | 0.509 | 0.061
15000 * 1000 | 0.616 | 0.079
20000 * 1000 | 0.828 | 0.115
10000 * 2000 | 0.831 | 0.101
12000 * 2000 | 0.988 | 0.120
15000 * 2000 | 1.235 | 0.151
20000 * 2000 | 1.685 | 0.201
10000 * 3000 | 1.306 | 0.169
12000 * 3000 | 1.579 | 0.204
15000 * 3000 | 1.864 | 0.224
20000 * 3000 | 2.522 | 0.312

<details>
<summary>script</summary>

```python

import numpy as np
import cupy as cp
import cupyx

n_repeat = 2
n_warmup = 1
crng = cp.random.default_rng()
nrng = np.random.default_rng()

def get_time(perf):
    cpu_time = perf.cpu_times.mean()
    gpu_time = perf.gpu_times.mean()
    return max(cpu_time, gpu_time)

def run(p, q):
    times = []

    perf = cupyx.time.repeat(nrng.power, (p,),
                            n_repeat=n_repeat, n_warmup=n_warmup)
    times.append(get_time(perf))
    perf = cupyx.time.repeat(crng.power, (q,),
                            n_repeat=n_repeat, n_warmup=n_warmup)
    times.append(get_time(perf))
    return times


print('size(nrows * ncols) | numpy (sec) | cupyx (sec)')
print('---|---|---')
nrows = [10000, 12000, 15000, 20000]
ncols = [1, 1000, 2000, 3000]
for m in ncols:
    for n in nrows:
        p = np.random.random((n, m))
        q = cp.asarray(p)
        times = run(p, q)
        print('{} * {} | {:.3f} | {:.3f}'.format(n, m, times[0], times[1]))

### CODE REVIEW

File: cupy/random/cupy_distributions.cu [hunk 1]
```diff
@@ -489,6 +489,13 @@ struct standard_gamma_functor {
     }
 };
 
+struct power_functor {
```
Comment (emcastillo) [inline cupy/random/cupy_distributions.cu:492]: There is already a `standard_exponential_functor` no need to alias it

Comment (emcastillo) [reply cupy/random/cupy_distributions.cu:492]: Also you are not using this one 😄

File: cupy/random/cupy_distributions.cu [hunk 2]
```diff
@@ -489,6 +489,13 @@ struct standard_gamma_functor {
...
+    template<typename... Args>
+    __device__ double operator () (Args&&... args) {
+        return rk_standard_exponential(args...);
+    }
+};
+
```
Comment (emcastillo) [inline cupy/random/cupy_distributions.cu:498]:
```suggestion
```

File: cupy/random/cupy_distributions.cu [hunk 3]
```diff
@@ -583,6 +590,11 @@ void poisson(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t
     generator_dispatcher(generator, launcher, state, out, size, reinterpret_cast<int64_t*>(lam));
 }
 
+void power(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream) {
+    kernel_launcher<power_functor, double> launcher(size, reinterpret_cast<cudaStream_t>(stream));
+    generator_dispatcher(generator, launcher, state, out, size);
+}
+
```
Comment (emcastillo) [inline cupy/random/cupy_distributions.cu:597]:
```suggestion
```

File: cupy/random/cupy_distributions.cuh [hunk 4]
```diff
@@ -33,6 +33,7 @@ void exponential(int generator, intptr_t state, intptr_t out, ssize_t size, intp
 void geometric(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream, intptr_t p);
 void hypergeometric(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream, intptr_t ngood, intptr_t nbad, intptr_t nsample);
 void poisson(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream, intptr_t lam);
+void power(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream);
```
Comment (emcastillo) [inline cupy/random/cupy_distributions.cuh:36]:
```suggestion
```

File: cupy/random/cupy_distributions.cuh [hunk 5]
```diff
@@ -54,6 +55,7 @@ void exponential(int generator, intptr_t state, intptr_t out, ssize_t size, intp
 void geometric(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream, intptr_t p) {}
 void hypergeometric(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream, intptr_t ngood, intptr_t nbad, intptr_t nsample) {}
 void poisson(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream, intptr_t lam) {}
+void power(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream) {}
```
Comment (emcastillo) [inline cupy/random/cupy_distributions.cuh:58]:
```suggestion
```

File: cupy/random/_generator_api.pyx [hunk 6]
```diff
@@ -57,6 +57,9 @@ cdef extern from 'cupy_distributions.cuh' nogil:
     void poisson(
         int generator, intptr_t state, intptr_t out,
         ssize_t size, intptr_t stream, intptr_t arg1)
+    void power(
+        int generator, intptr_t state, intptr_t out,
+        ssize_t size, intptr_t stream)
```
Comment (emcastillo) [inline cupy/random/_generator_api.pyx:62]:
```suggestion
```

**Review by emcastillo — CHANGES_REQUESTED** [review]
You added a `power` function but you are not using it (Which is cool)
please remove it :)

File: tests/cupy_tests/random_tests/common_distributions.py [hunk 7]
```diff
@@ -338,3 +338,29 @@ def test_hypergeometric_ks(self):
                 or isinstance(self.nsample, list)):
             self.skipTest('Stastical checks only for scalar args')
         self.check_ks(0.05)(self.ngood, self.nbad, self.nsample, size=2000)
+
+
+power_params = [
+    {'a': 0.5},
+    {'a': 1},
+    {'a': 5},
+    {'a': [0.8, 0.7, 1, 2, 5]},
```
Comment (emcastillo) [inline tests/cupy_tests/random_tests/common_distributions.py:347]: won't this parameter be ignored by the below checks?

**Comment by emcastillo** [comment]
Jenkins, test this please

**Review by takagi — DISMISSED** [review]

**Comment by takagi** [comment]
Jenkins, test this please

**Review by takagi — APPROVED** [review]

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
LGTM!

**Comment by takagi** [comment]
@emcastillo would you approve the change?

**Review by emcastillo — APPROVED** [review]

**Pull request merged**